### ISSUE TITLE
Rule request: no-duplicate-assignment

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [palantir/tslint#2183]
I recently had a code review in which someone else made submitted similar, but not identical, changes to the file. Merging their changes in produced syntactically valid but stylistically silly TypeScript code:

```typescript
interface IContainerArgs {
    age: number;
    name: string;
}

class Container {
    private age: number;
    private name: string;

    constructor(args: IContainerArgs) {
        this.age = args.age;
        this.name = args.name;
        this.name = args.name;
        this.age = args.age;
    }
}
```

It would be nice to have a TSLint rule that informs the user they're doing unnecessary work. It could be just from a bad merge, or maybe something worse (unexpected behavior from an incorrect ordering?).

One could make the argument that a design could necessitate setting something twice, such as for an intended side effect of a getter or setter... but that feels like inherently confusing and poorly designed code.

#### Expected

```typescript
this.age = args.age;
this.name = args.name;
this.name = args.name;    // TSLint error: unnecessary duplicate set
this.age = args.age;    // TSLint error: unnecessary duplicate set
```

Note that the rule shouldn't flag operations after the thing being assigned to performs any operation.

```typescript
this.age = args.age;
this.useAgeAndMutateIt();
this.age = args.age; // fine
```

Whether "any operation" includes getters/setters could be up to a rule configuration, but I think it should exclude them by default. Getters and setters mutating things is weird.

```typescript
this.age = args.age;
this.accessAge;
this.age = args.age;    // Error by default?
```

**Comment by nchen63** [comment]
maybe `no-duplicate-assignment` would be more clear, since "set" makes me think of the set datatype or concept first

**Comment by JoshuaKGoldberg** [comment]
Note: per #4534, this issue will be closed in less than a month if no PR is sent to add the rule. If you _really_ need the rule, custom rules are always an option and can be maintained outside this repo!

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated** and no longer accepting pull requests for new rules. See #4534. 😱

If you'd like to see this rule implemented, you have two choices:
* **Recommended:** Use [typescript-eslint](https://typescript-eslint.io) to lint your TypeScript code with ESLint ✅
* **Legacy:** Write it as a [custom TSLint rule](https://palantir.github.io/tslint/develop/custom-rules/) 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
